test/unit/org/apache/cassandra/cql3/ViewTest.java

```diff
@@ -1469,4 +1460,32 @@ public void testQuotedIdentifiersInWhereClause() throws Throwable
                    row("\"theKey\" IS NOT NULL AND (\"theClustering_1\", \"theClustering_2\") = (1, 2) AND \"theValue\" IS NOT NULL"),
                    row("token(\"theKey\") > token(1) AND \"theClustering_1\" = 1 AND \"theClustering_2\" > 2 AND \"theValue\" IS NOT NULL"));
     }
+
+    /**
+     * Tests that truncating a table stops the ongoing builds of its materialized views,
+     * so they don't write into the MV data that has been truncated in the base table.
+     *
+     * See CASSANDRA-16567 for further details.
+     */
+    @Test
+    @BMRule(name = "Delay materialized view mutations",
+    targetClass = "StorageProxy",
+    targetMethod = "mutateMV",
+    action = "Thread.sleep(4000);")
+    public void testTruncateWhileBuilding() throws Throwable
+    {
+        createTable("CREATE TABLE %s (k int, c int, v int, PRIMARY KEY(k, c))");
+        execute("USE " + keyspace());
+        executeNet("USE " + keyspace());
+        updateView("INSERT INTO %s (k, c, v) VALUES (?, ?, ?)", 0, 0, 0);
+        createView("mv",
+                   "CREATE MATERIALIZED VIEW %s AS SELECT * FROM %%s " +
+                   "WHERE k IS NOT NULL AND c IS NOT NULL AND v IS NOT NULL " +
+                   "PRIMARY KEY (v, c, k)");
+        Thread.sleep(2000); // wait for the creation of MV build tasks
```

Comment:
I have replaced the sleeps by a `CountDownLatch` manipulated by a couple of Byteman rules. One rule blocks the view builder tasks on the latch, and the other rule unblocks them when the truncation starts. That seems enough to consistently reproduce the bug when we don't stop the build on `truncateBlocking`. I'll give it a few runs in the multiplexer just in case.

Reply:
Looks quite good, thanks